### PR TITLE
feature: damlc lints all files in project directory

### LINKED ISSUE #8887: daml damlc lint should support linting a whole project
Currently we accept only a single file which is pretty annoying. We should make the flie optional and default to linting the whole project.

### PR DESCRIPTION
**Pull request by robin-da**
fixes #8887.

This changes the input options of `damlc lint` to take several files. If
no file is given, all `.daml` files contained in the directory are
linted.

CHANGELOG_BEGIN
CHANGELOG_END

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: compiler/damlc/lib/DA/Cli/Damlc.hs [hunk 1]
```diff
@@ -539,25 +539,32 @@ execCompile inputFile outputFile opts (WriteInterface writeInterface) mbIfaceDir
...
+               setFilesOfInterest ide (HashSet.singleton inputFile)
+               runActionSync ide $ getDlintIdeas inputFile
+               diags <- getDiagnostics ide
+               if null diags then
+                 hPutStrLn stderr "No hints"
+               else
+                 exitFailure
+     getInputFiles = \case
+       [] -> do
+           fs <- listFilesInside (pure . not . isPrefixOf ".daml" . takeFileName) "."
```
Comment (cocreature) [inline compiler/damlc/lib/DA/Cli/Damlc.hs:565]: this doesn’t seem quite right. We only want files in the source directory from your daml.yaml. We can probably share the logic that `daml build` uses to find those.

Comment (robin-da) [reply compiler/damlc/lib/DA/Cli/Damlc.hs]: using `getDamlRootFiles` now.

File: compiler/damlc/lib/DA/Cli/Damlc.hs [hunk 2]
```diff
@@ -539,25 +539,31 @@ execCompile inputFile outputFile opts (WriteInterface writeInterface) mbIfaceDir
...
-         withDamlIdeState opts loggerH diagnosticsLogger $ \ide -> do
-             setFilesOfInterest ide (HashSet.singleton inputFile)
-             runActionSync ide $ getDlintIdeas inputFile
-             diags <- getDiagnostics ide
-             if null diags then
-               hPutStrLn stderr "No hints"
-             else
-               exitFailure
+         inputFiles <- getInputFiles relativize inputFiles
+         forM_ inputFiles $ \inputFile -> do
```
Comment (cocreature) [inline compiler/damlc/lib/DA/Cli/Damlc.hs:552]: can we move the loop into `withDamlIdeState` and only set files of interest once to all of them? I think that should be significantly faster. You probably also want to always lint all files and only print "No hints" once (or call `exitFailure`).

Comment (robin-da) [reply compiler/damlc/lib/DA/Cli/Damlc.hs]: moved it to the main loop

**Review by cocreature — APPROVED** [review]
Thanks! Would be great to have a test for this somewhere. Probably easiest to just create a new test suite.

**Comment by robin-da** [comment]
@cocreature cool, I'll add one.

**Pull request merged**